lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -137,6 +156,69 @@ void SPIRVRegularizeLLVM::lowerMemset(MemSetInst *MSI) {
   return;
 }
 
+void SPIRVRegularizeLLVM::buildFunnelShiftLeftFunc(Function *FSHLFunc) {
+  if (!FSHLFunc->empty())
+    return;
+
+  // Our function will require 3 basic blocks; the purpose of each will be
+  // clarified below.
+  auto *CondBB = BasicBlock::Create(M->getContext(), "cond", FSHLFunc);
+  auto *RotateBB =
+      BasicBlock::Create(M->getContext(), "rotate", FSHLFunc); // Main logic
+  auto *PhiBB = BasicBlock::Create(M->getContext(), "phi", FSHLFunc);
+
+  IRBuilder<> Builder(CondBB);
+  // If the number of bits to rotate for is divisible by the bitsize,
+  // the shift becomes useless, and we should bypass the main logic in that
+  // case.
+  auto *IntTy = cast<IntegerType>(FSHLFunc->getReturnType());
```

Comment:
I've just realized that intrinsic can accept vectors as arguments: in this case we need to apply desired transformation to each vector element.

If I understand correctly, this shouldn't be a huge problem, because each operation used supports vector as arguments, except `br`: we can't perform a jump based on vector of booleans. Having vectors in mind we probably need to either drop this optimization for the case when rotate is not needed at all or go for it only if the rotate is not needed for any vector element

Reply:
The vector catch is a really nice one. I'm also thinking it's not too much of a problem, so I'll add vector support as a separate commit.
As far as `br` is concerned, I'm leaning towards
> go for it only if the rotate is not needed for any vector element

Reply:
Done. Entirely dropping the condition-based optimization shouldn't cause too much problems, after all - it only affects a minority of cases for non-LLVM-based backends.